app/views/handlers/feedback.py

```diff
@@ -86,29 +89,49 @@ def handle_post(self) -> None:
+        metadata = self._questionnaire_store.metadata
+        case_id = metadata.case_id if metadata else None
+        tx_id = metadata.tx_id if metadata else None
+
+        feedback_metadata = FeedbackMetadata(case_id, tx_id)  # type: ignore
 
         # pylint: disable=no-member
         # wtforms Form parents are not discoverable in the 2.3.3 implementation
-        feedback_message = FeedbackPayload(
-            metadata=self._questionnaire_store.metadata,
-            response_metadata=self._questionnaire_store.response_metadata,
-            schema=self._schema,
-            case_id=self._questionnaire_store.metadata["case_id"],
-            submission_language_code=session_data.language_code,
-            feedback_count=session_data.feedback_count,
-            feedback_text=self.form.data.get("feedback-text"),
-            feedback_type=self.form.data.get("feedback-type"),
-        )
-        message = feedback_message()
-        metadata = feedback_metadata()
-        message.update(metadata)
+        # type ignore as metadata will exist at this point
+        if metadata.version == AuthPayloadVersion.V2.value:  # type: ignore
+            feedback_message = FeedbackPayloadV2(
+                metadata=metadata,  # type: ignore
+                response_metadata=self._questionnaire_store.response_metadata,
+                schema=self._schema,
+                case_id=case_id,
+                submission_language_code=session_data.language_code,
+                feedback_count=session_data.feedback_count,
+                feedback_text=self.form.data.get("feedback-text"),
+                feedback_type=self.form.data.get("feedback-type"),
+            )
+        else:
+            feedback_message = FeedbackPayload(
+                metadata=metadata,  # type: ignore
+                response_metadata=self._questionnaire_store.response_metadata,
+                schema=self._schema,
+                case_id=case_id,
+                submission_language_code=session_data.language_code,
+                feedback_count=session_data.feedback_count,
+                feedback_text=self.form.data.get("feedback-text"),
+                feedback_type=self.form.data.get("feedback-type"),
+            )
+
+        feedback_message().update(feedback_metadata())
         encrypted_message = encrypt(
-            message, current_app.eq["key_store"], KEY_PURPOSE_SUBMISSION  # type: ignore
+            feedback_message(), current_app.eq["key_store"], KEY_PURPOSE_SUBMISSION  # type: ignore
         )
 
+        if metadata.version == AuthPayloadVersion.V2.value and metadata.survey_metadata.receipting_keys:  # type: ignore
+            receipting_keys: dict = {item: metadata[item] for item in metadata.survey_metadata.receipting_keys}  # type: ignore
+            feedback_metadata = FeedbackMetadata(case_id, tx_id, **receipting_keys)  # type: ignore
+
```

Comment:
We already have feedback metadata, but we then regenerate it which is a bit odd and unclear. I would remove the `update` call. This is added metadata to feedback payload but case id is already passed it, so all that is doing is adding `tx_id`. Therefore. I would remove L124 and just pass in tx_id to `FeedbackPayload`.

Then metadata can just be generated. I've also added a check to ensure survey metadata exists. The type ignore is hiding the error, hence why we need to be careful with type ignores.


```py
        additional_metadata = {}
        if (
            metadata.version is AuthPayloadVersion.V2
            and metadata.survey_metadata
            and metadata.survey_metadata.receipting_keys
        ):
            additional_metadata: dict = {
                item: metadata[item]
                for item in metadata.survey_metadata.receipting_keys
            }

        feedback_metadata = FeedbackMetadata(
            tx_id=metadata.tx_id, case_id=metadata.case_id, **additional_metadata
        )
```

Reply:
Updated 👍